### PR TITLE
Html parsing error on repeated tags sf205

### PR DESCRIPTION
**Pull request by damien-rembert**
## Pull request type

- Bug fix -> [bug]

## Which ticket is resolved?

- script + style HTML elements in source, missing stylesheet in result.
  * URL: [SourceForge bug 609](https://sourceforge.net/p/omegat/bugs/609/) 

- HTML parsing error
  * URL: [SourceForge bug 205](https://sourceforge.net/p/omegat/bugs/205/)

## What does this PR change?

-Split the recurse boolean
-Add missing HTML elements that can contain text and act as start of paragraph
-Rename variables to make the process of FilterVisitor clearer
-Correct a typo in the PR template
-Add annoying MacOS system files to gitignore

### CODE REVIEW

File: src/org/omegat/filters2/html2/FilterVisitor.java [hunk 1]
```diff
@@ -82,11 +84,16 @@ public FilterVisitor(HTMLFilter2 htmlfilter, BufferedWriter bufwriter, HTMLOptio
     // Variable declaration
     // ///////////////////////////////////////////////////////////////////////
 
-    /** Should the parser call us for this tag's ending tag and its inner tags. */
-    protected boolean recurse = true;
+    /** Should the parser call us for this tag's ending tag. */
```
Comment (damien-rembert) [inline src/org/omegat/filters2/html2/FilterVisitor.java:87]: This change fixes the bug. Most of the other changes are renaming variables to make the code clearer.

File: src/org/omegat/filters2/html2/FilterVisitor.java [hunk 2]
```diff
@@ -153,37 +161,21 @@ public boolean shouldRecurseChildren() {
...
-                Attribute attribute = (Attribute) i.next();
-                String name = attribute.getName();
-                String value = attribute.getValue();
-                if (name == null || value == null) {
-                    continue;
-                }
-                keepIntact = this.filter.checkIgnoreTags(name, value);
-            }
-        }
+        boolean keepIntact = isProtectedTag(tag);
```
Comment (damien-rembert) [inline src/org/omegat/filters2/html2/FilterVisitor.java:164]: isProtected() now covers the behaviour deleted above

**Comment by brandelune** [comment]
Do you have test HTML files ?

**Comment by damien-rembert** [comment]
I have been using these files (I think they were more or less inspired by the ones provided in the bug reports): 

https://github.com/omegat-org/omegat/blob/ed78e6ce9cb2a871aa264d690b6eed3eac1ed72b/test/data/filters/html/file-HTMLFilter2-all-block-elements.html

https://github.com/damien-rembert/omegat/blob/HTML-parsing-error-on-repeated-tags-SF205/test/data/filters/html/file-HTMLFilter2-recurse-bugfix.html

File: src/org/omegat/filters2/html2/FilterVisitor.java [hunk 3]
```diff
@@ -82,11 +84,16 @@ public FilterVisitor(HTMLFilter2 htmlfilter, BufferedWriter bufwriter, HTMLOptio
...
 
-    /** Should the parser call us for this tag's ending tag and its inner tags. */
-    protected boolean recurse = true;
+    /** Should the parser call us for this tag's ending tag. */
+    protected boolean recurseSelf = true;
+
+    /**
+     * Should the parser call us for this tag's inner tags.
+     */
+    protected boolean recurseChildren = true;
```
[the remark this answers is not in the document]
Comment (brandelune) [reply src/org/omegat/filters2/html2/FilterVisitor.java:94]: I don't have a complex HTML file set at hand to test, so if the test files produce the expected results, I'm fine with it.

File: src/org/omegat/filters2/html2/HTMLFilter2.java [hunk 4]
```diff
@@ -179,7 +179,7 @@ public void processFile(BufferedReader infile, BufferedWriter outfile,
             parser.setInputHTML(all.toString());
             parser.visitAllNodesWith(new FilterVisitor(this, outfile, options));
         } catch (ParserException pe) {
-            System.out.println(pe);
+            Log.log(pe);
```
Comment (miurahr) [inline src/org/omegat/filters2/html2/HTMLFilter2.java:182]: it would be better to be `Log.logErrorRB(pe, "<bundle key>");

Comment (damien-rembert) [reply src/org/omegat/filters2/html2/HTMLFilter2.java]: Thanks, edited accordingly

**Comment by damien-rembert** [comment]
I'll have a look at the failing checks tomorrow

**Comment by damien-rembert** [comment]
Sorry, it needs more work :')

File: test/data/filters/html/file-HTMLFilter2-tag-dropping-bugfix-SF609.html [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<!doctype html>
+<html>
+<head><meta charset="UTF-8"><script>{}</script><style>
```
Comment (damien-rembert) [inline test/data/filters/html/file-HTMLFilter2-tag-dropping-bugfix-SF609.html:3]: The failing test we saw before seems to have been caused by the case of `utf-8`: OmegaT turns it into upper-case letters which caused the assert to fail. This behaviour (the change of case) existed before and is not related to the bugs as far as I can tell.

File: src/org/omegat/filters2/html2/FilterVisitor.java [hunk 6]
```diff
@@ -389,6 +421,7 @@ private boolean isProtectedTag(Tag tag) {
         String tagname = tag.getTagName();
 
         String[] noEditTags = { "!DOCTYPE", "STYLE", "SCRIPT", "OBJECT", "EMBED" };
+        // String[] noEditTags = { "!DOCTYPE", "SCRIPT", "OBJECT", "EMBED" };
```
Comment (miurahr) [inline src/org/omegat/filters2/html2/FilterVisitor.java]: What this?

File: src/org/omegat/filters2/html2/FilterVisitor.java [hunk 7]
```diff
@@ -206,9 +234,13 @@ public void visitTag(Tag tag) {
...
                                 && ("submit".equalsIgnoreCase(tag.getAttribute("type"))
                                         || "button".equalsIgnoreCase(tag.getAttribute("type"))
-                                        || "reset".equalsIgnoreCase(tag.getAttribute("type"))
-                           ) //and it is a submit/button/reset element.
-                   ) {
+                                        || "reset".equalsIgnoreCase(tag.getAttribute("type"))) // and
+                                                                                               // it
+                                                                                               // is
+                                                                                               // a
+                                                                                               // submit/button/reset
```
Comment (miurahr) [inline src/org/omegat/filters2/html2/FilterVisitor.java:213]: It is better to improve comment line

File: src/org/omegat/filters2/html2/FilterVisitor.java [hunk 8]
```diff
@@ -571,6 +604,8 @@ protected void endup() {
         for (int i = 0; i < firstTagToIncludeFromPreceding; i++) {
             Node node = allNodesInParagraph.get(i);
             if (node instanceof Tag) {
+                // Tag nodeAsATag = (Tag) node;
+                // writeout(nodeAsATag.toHtml());
```
Comment (miurahr) [inline src/org/omegat/filters2/html2/FilterVisitor.java]: Why leave a debug lines(?) in comments?

File: src/org/omegat/filters2/html2/FilterVisitor.java [hunk 9]
```diff
@@ -163,12 +163,40 @@ public void visitTag(Tag tag) {
...
+        /** Do we collect the translatable text now. */
+        Log.log("isTextUpForCollection " + isTextUpForCollection);// = false;
+
+        /**
+         * Did the PRE block start (it means we mustn't compress the spaces).
+         */
+        Log.log("preformatting " + preformatting);// = false;
+        Log.log("hasAnEndTag " + hasAnEndTag(tag));// = false;
+
+        // }
```
Comment (miurahr) [inline src/org/omegat/filters2/html2/FilterVisitor.java]: looks like debug messages are written out as INFO category.
It should be `Log.debug()` or somethings similar.

**Review by miurahr — CHANGES_REQUESTED** [review]
There are several debug lines left in live.
Please clean-up or properly set debug category for log.

**Comment by damien-rembert** [comment]
Sorry, it seems like I pushed changes at about the same time you were reviewing :)
Hopefully most of these are fixed now.
I'll double-check later today.

**Comment by damien-rembert** [comment]
@miurahr Sorry you had to go through my investigation commit, I'm not sure why GitHub decided to show you these changes and why they're still not marked as outdated 🤷 
I will make the other changes you requested later today :)

**Comment by miurahr** [comment]
There are 15 commits that seems not identical each commits.
Please recreate commit history that is valuable for merge?

I reviewed commits by commits to understand what commit doing.
When it is not finished the work, please set [WIP] or set as draft PR.

**Comment by damien-rembert** [comment]
> There are 15 commits that seems not identical each commits.
> Please recreate commit history that is valuable for merge?

Sure, I'll do that later today.

> When it is not finished the work, please set [WIP] or set as draft PR.

If you look at the history of the PR, it was marked as draft as soon as I realised there was a problem with the test and the PR stayed there until it was fixed, I don't know why it wasn't displayed properly during your review.

**Review by miurahr — APPROVED** [review]
A test coverage of FilterVisitor class is 91% in instructions, and 76% branch, that is good enough for OmegaT average.

**Comment by miurahr** [comment]
> A test coverage of FilterVisitor class is 91% in instructions, and 76% branch, that is good enough for OmegaT average.

Especially, a chunk of main changes are covered. LGTM.

**Comment by damien-rembert** [comment]
This can be closed as it was included in #452

**Pull request closed without merging**